src/envoy/mixer/tcp_filter.cc

```diff
@@ -0,0 +1,206 @@
+      Ssl::Connection* ssl = filter_callbacks_->connection().ssl();
+      if (ssl != nullptr) {
+        origin_user = ssl->uriSanPeerCertificate();
+      }
+
+      filter_callbacks_->connection().readDisable(true);
+      auto instance = GetPtr();
+      calling_check_ = true;
+      mixer_control_->CheckTcp(request_data_, filter_callbacks_->connection(),
+                               origin_user, [instance](const Status& status) {
+                                 instance->completeCheck(status);
+                               });
+      calling_check_ = false;
+    }
+    return state_ == State::Calling ? Network::FilterStatus::StopIteration
+                                    : Network::FilterStatus::Continue;
+  }
+
+  void completeCheck(const Status& status) {
+    log().debug("Called TcpInstance completeCheck: {}", status.ToString());
+    if (state_ == State::Closed) {
+      return;
+    }
+    state_ = State::Completed;
+    filter_callbacks_->connection().readDisable(false);
+
+    if (!status.ok()) {
+      check_status_code_ = status.error_code();
+      filter_callbacks_->connection().close(
+          Network::ConnectionCloseType::NoFlush);
+    } else {
+      if (!calling_check_) {
+        filter_callbacks_->continueReading();
+      }
+    }
+  }
+
+  // Network::ConnectionCallbacks
+  void onEvent(uint32_t events) override {
+    if (filter_callbacks_->upstreamHost()) {
+      log().debug("Called TcpInstance onEvent: {} upstream {}", events,
+                  filter_callbacks_->upstreamHost()->address()->asString());
+    } else {
+      log().debug("Called TcpInstance onEvent: {}", events);
+    }
+
+    if (events & Network::ConnectionEvent::RemoteClose ||
+        events & Network::ConnectionEvent::LocalClose) {
+      if (state_ != State::Closed && request_data_) {
+        mixer_control_->ReportTcp(request_data_, request_bytes_,
+                                  response_bytes_, check_status_code_,
+                                  filter_callbacks_->upstreamHost());
+      }
+      state_ = State::Closed;
+    }
+  }
+};
+
+}  // namespace Mixer
+}  // namespace Http
+
+namespace Server {
+namespace Configuration {
+
+class TcpMixerFilter : public NetworkFilterConfigFactory {
+ public:
+  NetworkFilterFactoryCb tryCreateFilterFactory(
+      NetworkFilterType type, const std::string& name,
+      const Json::Object& config, Server::Instance& server) override {
+    if (type != NetworkFilterType::Both || name != "mixer") {
+      return nullptr;
+    }
+
+    Http::Mixer::TcpConfigPtr tcp_config(
+        new Http::Mixer::TcpConfig(config, server));
+    return [tcp_config](Network::FilterManager& filter_manager) -> void {
+      std::shared_ptr<Http::Mixer::TcpInstance> instance =
+          std::make_shared<Http::Mixer::TcpInstance>(tcp_config);
+      filter_manager.addReadFilter(Network::ReadFilterSharedPtr(instance));
+      filter_manager.addWriteFilter(Network::WriteFilterSharedPtr(instance));
```

Comment:
It seems that it is the only way to get write_bytes.   Other approaches may require Envoy core code change.